src/main/scala/firrtl/passes/LowerTypes.scala

```diff
@@ -81,165 +78,145 @@ object LowerTypes extends Pass {
+      case "addr" | "en" | "clk" | "wmode" =>
+        require(tail.isEmpty) // there can't be a tail for these
+        memDataTypeMap(mem.name) match {
+          case _: GroundType => Seq(e)
+          case memType => create_exps(mem.name, memType) map { e =>
+            val loMemName = loweredName(e)
+            val loMem = WRef(loMemName, UnknownType, kind(mem), UNKNOWNGENDER)
+            mergeRef(loMem, mergeRef(port, field))
+          }
         }
-      }
+      // Fields that need not be replicated for each
+      // eg. mem.reader.data[0].a
+      // (Connect/IsInvalid must already have been split to ground types)
+      case "data" | "mask" | "rdata" | "wdata" | "wmask" =>
+        val loMem = tail match {
+          case Some(e) =>
+            val loMemExp = mergeRef(mem, e)
+            val loMemName = loweredName(loMemExp)
+            WRef(loMemName, UnknownType, kind(mem), UNKNOWNGENDER)
+          case None => mem
+        }
+        Seq(mergeRef(loMem, mergeRef(port, field)))
+      case name => error(s"Error! Unhandled memory field ${name}")(info, mname)
+    }
+  }
 
-      def lowerTypesExp(e: Expression): Expression = e match {
-        case e: WRef => e
-        case (_: WSubField | _: WSubIndex) => kind(e) match {
-          case k: InstanceKind =>
-            val (root, tail) = splitRef(e)
-            val name = loweredName(tail)
-            WSubField(root, name, e.tpe, gender(e))
-          case k: MemKind =>
-            val exps = lowerTypesMemExp(e)
-            exps.size match {
-              case 1 => exps.head
-              case _ => error("Error! lowerTypesExp called on MemKind " + 
-                              "SubField that needs to be expanded!")
-            }
-          case _ => WRef(loweredName(e), e.tpe, kind(e), gender(e))
+  def lowerTypesExp(memDataTypeMap: MemDataTypeMap,
+      info: Info, mname: String)(e: Expression): Expression = e match {
+    case (_: WSubField | _: WSubIndex) => kind(e) match {
+      case k: InstanceKind =>
+        val (root, tail) = splitRef(e)
+        val name = loweredName(tail)
+        WSubField(root, name, e.tpe, gender(e))
+      case k: MemKind =>
+        val exps = lowerTypesMemExp(memDataTypeMap, info, mname)(e)
+        exps.size match {
+          case 1 => exps.head
+          case _ => error("Error! lowerTypesExp called on MemKind " + 
+                          "SubField that needs to be expanded!")(info, mname)
         }
-        case e: Mux => e map (lowerTypesExp)
-        case e: ValidIf => e map (lowerTypesExp)
-        case e: DoPrim => e map (lowerTypesExp)
-        case e @ (_: UIntLiteral | _: SIntLiteral) => e
-      }
+      case _ => WRef(loweredName(e), e.tpe, kind(e), gender(e))
+    }
+    case e: Mux => e map lowerTypesExp(memDataTypeMap, info, mname)
+    case e: ValidIf => e map lowerTypesExp(memDataTypeMap, info, mname)
+    case e: DoPrim => e map lowerTypesExp(memDataTypeMap, info, mname)
+    case _ => e
```

Comment:
Don't do this, match on WRef specifically, and for the case _ => ..., throw an error. We want to error the compiler if we see a WSubAccess